### PR TITLE
Changes for task 40

### PR DESCRIPTION
**Pull request by dulasevich**

### CODE REVIEW

File: src/main/java/by/Variables.java [hunk 1]
```diff
@@ -11,4 +11,26 @@ public class Variables {
     public static By forwardEmailIcon = By.xpath("//div[contains(@data-key, 'forward')]");
     public static By pageTitle = By.tagName("title");
     public static By currentDate = By.cssSelector("span[class$='ToggleWidget-DayOfMonth']");
+
+    //task 40 new locators
+    public static By accountName = By.xpath("//div[contains(@class, 'legouser')]/a/span[@class='user-account__name']");
```
Comment (LikBez123) [inline src/main/java/by/Variables.java]: make it final and in uppercase

File: src/main/java/by/Variables.java [hunk 2]
```diff
@@ -11,4 +11,26 @@ public class Variables {
...
+    public static By mailIcon = By.xpath("//span[text()='Почта' or text()='Mail']");
+    public static By alert = By.xpath("//button[@onclick='myAlertFunction()']");
+    public static By js = By.xpath("//button[@onclick='myConfirmFunction()']");
```
Comment (LikBez123) [inline src/main/java/by/Variables.java]: please check the names of elements

File: src/test/java/AlertsTest.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        driver = new ChromeDriver();
+        driver.manage().window().maximize();
+        driver.get("https://demo.seleniumeasy.com/javascript-alert-box-demo.html");
+    }
+
+    @Test
+    void javaScriptAlertTest() {
+        WebElement alertBox = driver.findElement(alert);
+        alertBox.click();
+
```
Comment (LikBez123) [inline src/test/java/AlertsTest.java]: remove empty lines

File: src/test/java/AlertsTest.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        Assertions.assertTrue(driver.findElement(jsResult).getText().equals("You pressed OK!"));
+    }
+
+    @Test
+    void javaScriptAlertPromptTest() {
+        final String sendKeys = "Dmitry";
+        WebElement promptBox = driver.findElement(prompt);
+        promptBox.click();
+
+        driver.switchTo().alert().sendKeys(sendKeys);
```
Comment (LikBez123) [inline src/test/java/AlertsTest.java]: `Alert alert = driver.switchTo().alert();`

File: src/test/java/DownloadTest.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        driver.manage().window().maximize();
+        driver.get("https://demo.seleniumeasy.com/bootstrap-download-progress-demo.html");
+    }
+
+    @Test
+    void downloadTest() {
+        WebElement clickDownload = driver.findElement(By.cssSelector("#cricle-btn"));
+        clickDownload.click();
+
+        new WebDriverWait(driver, Duration.ofSeconds(15)).
```
Comment (LikBez123) [inline src/test/java/DownloadTest.java]: add polling every 100 mills
`Duration.ofMillis(100)`

File: src/test/java/LoginTest.java [hunk 6]
```diff
@@ -22,23 +21,32 @@ public class YandexTest {
...
+        Thread.sleep(1000);
+
         WebElement logIntoExistingAccount = driver.findElement(loginButton);
         logIntoExistingAccount.click();
 
         WebElement login = driver.findElement(loginField);
         login.sendKeys(LOGIN);
         login.submit();
 
+        Thread.sleep(1000);
```
Comment (LikBez123) [inline src/test/java/LoginTest.java]: remove redundant` Thread.sleep()-s`
use `implicitlyWait(Duration.ofSeconds(10));`

File: src/test/java/MultiSelectTest.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+    @Test
+    void multiTest() {
+        WebElement multi = driver.findElement(states);
+        Select select = new Select(multi);
+
+        List<String> allOptions = select.getOptions().stream().map(el -> el.getText()).collect(Collectors.toList());
+
+        Random r = new Random();
+        while (select.getAllSelectedOptions().size() != 3) {
```
Comment (LikBez123) [inline src/test/java/MultiSelectTest.java]: you can move this logic to separate method

File: src/test/java/ParamsTest.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        WebDriverManager.getInstance(ChromeDriver.class).setup();
+        driver = new ChromeDriver();
+        driver.manage().window().maximize();
+        driver.manage().timeouts().implicitlyWait(Duration.ofSeconds(3));
+        driver.get("https://mail.yandex.com/");
+    }
+
+    @ParameterizedTest()
+    @CsvSource({CREDENTIALS_1, CREDENTIALS_2})
+    void login (String l, String p){
```
Comment (LikBez123) [inline src/test/java/ParamsTest.java]: use correct parameter names

File: src/main/java/table/User.java [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+public class User {
+
+    private String name;
+    private String position;
+    private String office;
+    private int age;
+    private int salary;
+
+    public User(String name, String position, String office, int age, int salary) {
```
Comment (LikBez123) [inline src/main/java/table/User.java]: `User(String name, String position, String office)`

File: src/test/java/TableUserTest.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    @BeforeEach
+    void initializeBrowser() {
+        WebDriverManager.getInstance(ChromeDriver.class).setup();
+        driver = new ChromeDriver();
+        driver.manage().window().maximize();
+        driver.get("https://demo.seleniumeasy.com/table-sort-search-demo.html");
+    }
+
+    private List<User> filterUsers() {
+        List<WebElement> names = driver.findElements(userName);
```
Comment (LikBez123) [inline src/test/java/TableUserTest.java]:
```
@Test
    void tableTest() {
        TablePage tablePage = new TablePage(driver);
        tablePage.navigateToBaseURL();
        tablePage.selectOptionFromDropDown("10");
        List<Person> allPeopleWithAgeAndSalary = tablePage.getUsersWithConditions(30, 200000);
        for (int i = 0; i <= allPeopleWithAgeAndSalary.size() - 1; i++) {
            Person index = allPeopleWithAgeAndSalary.get(i);
            System.out.println(index);
        }
    }
```

```
public List<Person> getUsersWithConditions(int minAge, int maxSalary) {
        List<Person> persons = new ArrayList<>();
        do {
            WebElement nextButton = driver.findElement(By.cssSelector(".next"));
            collectUsersWithConditions(persons, minAge, maxSalary);
            nextButton.click();
        } while (driver.findElement(By.cssSelector(".next")).getAttribute("class").contains("disabled"));
        return persons;
    }

    private void collectUsersWithConditions(List<Person> users, int minAge, int maxSalary) {
        List<WebElement> rows = driver.findElements(By.cssSelector("tbody tr"));
        rows.forEach(row -> {
            int salary = Integer.parseInt(row.findElement(By.xpath("td[6]")).getAttribute("data-order"));
            int age = Integer.parseInt(row.findElement(By.xpath("td[4]")).getText());
            if (salary <= maxSalary && age > minAge) {
                String name = row.findElement(By.xpath("td[1]")).getText();
                String position = row.findElement(By.xpath("td[2]")).getText();
                String office = row.findElement(By.xpath("td[3]")).getText();
                users.add(new Person(name, position, office));
            }
        });
    }
```

**Review by LikBez123 — CHANGES_REQUESTED** [review]

**Comment by dulasevich** [comment]
For the latest commit - 'Changes for task 40 - updates per request' I have small question about TableUserTest, the solution that you provided with table rows and Next button.

File: src/test/java/MultiSelectTest.java [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+    @BeforeEach
+    void initializeDriver() {
+        WebDriverManager.getInstance(ChromeDriver.class).setup();
+        driver = new ChromeDriver();
+        driver.manage().window().maximize();
+        driver.get("https://demo.seleniumeasy.com/basic-select-dropdown-demo.html");
+    }
+
+    private Select randomThreeStatesSelection() {
```
Comment (LikBez123) [inline src/test/java/MultiSelectTest.java]: according to the java convention, it is better to call this method starting with the verb. Ex.: `selectStates()`, `selectThreeRandomStates()`

File: src/test/java/MultiSelectTest.java [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+        Random r = new Random();
+        while (select.getAllSelectedOptions().size() != 3) {
+            select.selectByIndex(r.nextInt(select.getOptions().size() - 1));
+        }
+        return select;
+    }
+
+    @Test
+    void multiTest() {
+        List<String> allOptions = randomThreeStatesSelection().getOptions().stream()
```
Comment (LikBez123) [inline src/test/java/MultiSelectTest.java]: it is better to define select in the test and not use the `randomThreeStatesSelection()` method to get an instance of the Select class

File: src/main/java/by/Variables.java [hunk 13]
```diff
@@ -3,12 +3,33 @@
...
+    public final static By PROMPT_ALERT_BOX = By.xpath("//button[@onclick='myPromptFunction()']");
+    public final static By JS_RESULT = By.xpath("//p[@id='confirm-demo']");
+    public final static By PROMPT_RESULT = By.xpath("//p[@id='prompt-demo']");
+    public final static By USER_IMAGE = By.xpath("//*[contains(@src, 'random')]");
+    public final static By USER_FIRST_NAME = By.xpath("//div[contains(text(), 'First Name')]");
+    public final static By PERCENT_ICON = By.xpath("//div[@class='percenttext']");
+    public final static By TABLE_ROWS = By.xpath("//tbody/tr");
+    public final static By NEXT_PAGE_ICON = By.id("example_next");
+    public final static By TABLE_NAME = By.xpath("td[1]");
+    public final static By TABLE_POSITION = By.xpath("td[2]");
```
Comment (LikBez123) [inline src/main/java/by/Variables.java]: .//td[2]

File: src/test/java/MultiSelectTest.java [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+                .map(WebElement::getText).collect(Collectors.toList());
+        List<String> selectedOptions = randomThreeStatesSelection().getAllSelectedOptions().stream()
+                .map(WebElement::getText).collect(Collectors.toList());
+        Assertions.assertTrue(allOptions.containsAll(selectedOptions));
```
Comment (LikBez123) [inline src/test/java/MultiSelectTest.java]: please check the test logic

**Review by LikBez123 — CHANGES_REQUESTED** [review]

**Comment by dulasevich** [comment]
написал не через стримы, чтобы не просто копировать твое решение, которое показал вчера вечером, ну и через стримы пока сложнее дается) 
.//td[2] для Variables сработало, но я бы все равно обсудил в след раз на всякий случай)

File: src/test/java/MultiSelectTest.java [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+    void initializeDriver() {
+        WebDriverManager.getInstance(ChromeDriver.class).setup();
+        driver = new ChromeDriver();
+        driver.manage().window().maximize();
+        driver.get("https://demo.seleniumeasy.com/basic-select-dropdown-demo.html");
+    }
+
+    private List<String> randomThreeStatesSelection(Select select) {
+        Random random = new Random();
+        select.getOptions();
```
Comment (LikBez123) [inline src/test/java/MultiSelectTest.java]: redundant line. Remove it

File: src/test/java/MultiSelectTest.java [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+        List<WebElement> allStates = select.getOptions();
+        List<String> expectedStates = new ArrayList<>();
+        int numberOfOptionsToBeSelected=3;
+        while (expectedStates.size() != numberOfOptionsToBeSelected) {
+            for (int i = 0; i < numberOfOptionsToBeSelected; i++) {
+                int randomIndex = random.nextInt(allStates.size());
+                expectedStates.add(allStates.get(randomIndex).getText());
+            }
+        }
+        return expectedStates;
```
Comment (LikBez123) [inline src/test/java/MultiSelectTest.java]: your logic is able to return a collection with the same elements. It is a mistake

**Review by LikBez123 — CHANGES_REQUESTED** [review]

**Review by LikBez123 — APPROVED** [review]

**Pull request merged**